src/dtx/dtx_common.c

```diff
@@ -784,56 +776,40 @@ dtx_batched_commit_deregister(struct ds_cont_child *cont)
 	int				 rc;
 
 	D_ASSERT(cont != NULL);
-	if (cont->sc_closing) {
-		D_ASSERT(cont->sc_dtx_flush_cbdata != NULL);
-
-		future = cont->sc_dtx_flush_cbdata;
-		goto wait;
-	}
+	D_ASSERT(cont->sc_open == 0);
 
 	head = &dss_get_module_info()->dmi_dtx_batched_list;
 	d_list_for_each_entry(dbca, head, dbca_link) {
 		if (uuid_compare(dbca->dbca_cont->sc_uuid,
 				 cont->sc_uuid) != 0)
 			continue;
 
-		if (--(dbca->dbca_shares) > 0)
-			return;
-
-		/* Notify the dtx_batched_commit ULT to flush the
-		 * committable DTXs via setting @sc_closing as 1.
+		/*
+		 * Notify the dtx_batched_commit ULT to flush the
+		 * committable DTXs.
 		 *
 		 * Then current ULT will wait here until the DTXs
 		 * have been committed by dtx_batched_commit ULT
 		 * that will wakeup current ULT.
 		 */
-		D_ASSERT(cont->sc_dtx_flush_cbdata == NULL);
-		D_ASSERT(cont->sc_dtx_flush_wait_count == 0);
-
+		D_ASSERT(dbca->dbca_deregistering == NULL);
 		rc = ABT_future_create(1, NULL, &future);
-		cont->sc_closing = 1;
 		if (rc != ABT_SUCCESS) {
 			D_ERROR("ABT_future_create failed for DTX flush on "
 				DF_UUID" %d\n", DP_UUID(cont->sc_uuid), rc);
 			return;
 		}
 
-		cont->sc_dtx_flush_cbdata = future;
-		goto wait;
-	}
-
-	D_ASSERT(0);
-
-wait:
-	cont->sc_dtx_flush_wait_count++;
-	rc = ABT_future_wait(future);
-	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed "
-		  "for DTX flush (2) on "DF_UUID": rc = %d\n",
-		  DP_UUID(cont->sc_uuid), rc);
+		dbca->dbca_deregistering = future;
+		rc = ABT_future_wait(future);
+		D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed "
```

Comment:
this function can't fail under any circumstance?

Reply:
I think it's not a good assertion (and the assertion on ABT_future_set() in batched commit ULT), but I didn't think of a good way to handle this kind of error so far. Fanyong, any thoughts?

Reply:
I agree that "D_ASSERTF(rc == ABT_SUCCESS)" is not good logic for checking ULT_wait return value, there is no guarantee that the ULT_wait will never fail. But if we really hit ULT_wait failure, then we will be in dilemmatic status: we do not know remote replica status, and we have no way to force abort/commit the DTX on remote replica. So here, either ignore the ULT_wait return value (as we do in some other use cases), or keep current assert() until we really trigger it in the future.